### PR TITLE
Travis CI configuration

### PR DESCRIPTION
**Pull request by strub**

### CODE REVIEW

**Comment by maximedenes** [comment]
Wonderful! You're making my life much easier :)

**Comment by matej-kosik** [comment]
I wonder, whether it would make sense to merge this as low as `v8.5` (and then all the way up)?

**Comment by silene** [comment]
Dumb question: why use trusty? If you use a more recent version, then you don't even need a ppa, do you?

**Comment by strub** [comment]
@silene, I used `trusty` because this is the last LTS... or used to think. Ok, going to move to `precise`.

@matej-kosik It is unclear if you want the same version of `OCaml` in the v8.5 branch. For trunk, we're forcing OCaml 4.02.3.

**Comment by strub** [comment]
@silena, I cannot find `opam` in `precise`... so, while I moved to `precise`, I'm going to continue using `ppa`.

**Comment by maximedenes** [comment]
Squashed and merged, thanks a lot!

**Comment by ejgallego** [comment]
This is great!! @strub will this build CoqIDE?

**Comment by JasonGross** [comment]
Is there any reason not to do:
```
addons:
  apt:
    sources:
    - avsm
  packages:
    - opam
```
and then not require sudo?

**Comment by ejgallego** [comment]
Also I dunno if using the system's Ocaml could bring some speed benefits.

**Comment by strub** [comment]
Ok, I'm going to add CoqIde and remove Audi (Travis documentation is inconsistent regarding this...)

Regarding using the system OCaml, we want to use an explicit version of OCaml.

**Comment by ejgallego** [comment]
Yeah I agree, maybe you could indeed add 3 builds, (4.02.3, 4.03, 4.04) ? AFAIK travis has support to run this in parallel.

**Comment by ejgallego** [comment]
I got coqide to work with this https://github.com/ejgallego/coq/commit/1a76ec34dd57489189a62f131f65d92a6d047e3c

Still documentation is not built, I dunno if that would be useful.

**Comment by ejgallego** [comment]
doc is hard to build as hevea conflicts with the ocaml packages in the ppa, we will have to disable this for now I'm afraid.

**Comment by ejgallego** [comment]
[edit: hevea is in opam, that may do the trick]

**Comment by Zimmi48** [comment]
Can someone explain where this is coming from, what is the purpose of this PR, why this is not redundant with Jenkins, and possibly why merge it so fast? I don't recall any discussion at a WG or on coqdev. But maybe I've just missed some stuff.

**Comment by maximedenes** [comment]
@Zimmi48 The fact that this is a PR and part of the codebase is just an artifact of github. It does not add code. Travis could as well have been configurable through external settings. We typically don't have a review process for our infrastructure. So maybe I should ask: why are you expecting a discussion, when e.g. Jenkins settings are usually not discussed?

I'm fine with changing our processes if people want to, but I think what was done here is consistent with the current processes.

**Comment by ejgallego** [comment]
Ok, that seems to work fine, https://github.com/ejgallego/coq/commit/072954ea731a83c5936c65c66634ee16bd270245 merge if you want.

@Zimmi48 this has been discussed many times, and this patch is indeed a life-saver for many people like me who have many trees and cannot use jenkins (despite our repeated requests to find a solution to it). Maxime is IMHO correct, no need to delay this as it is not related to Coq itself but to GH infrastructure.

**Comment by Zimmi48** [comment]
I didn't mean to say you did wrong. And by not discussed I meant: I have no clue what this is bringing because no one told me. Maybe Jenkins settings are not generally discussed but at least the purpose is well understood by everyone, I think.

**Comment by Zimmi48** [comment]
@ejgallego: do you have a pointer to those discussions or could you summarize what this brings? Why can't you use Jenkins?

**Comment by strub** [comment]
@ejgallego, would you mind incorporating Jason remarks ? I think he made a good point !

**Comment by ejgallego** [comment]
@strub, ok will take care of it.

@Zimmi48 AFAICT we have discussed the CI thing at every WG since I've been attending. I don't know why I can't use Jenkins, but the fact is I cannot (hint: how can I create an account?)

**Comment by maximedenes** [comment]
We want early automated testing of PRs, and Github webhooks for Jenkins are a nightmare. In fact, this is what I tried at the first coq coding sprint. I could make up a full report, discuss it at length and so on, but personally I'd prefer to keep the focus on coq-related issues.

**Comment by Zimmi48** [comment]
@maximedenes: Thanks, that was the kind of answer I was looking for. So is it going to supplant  Jenkins or is there a purpose for having both?

**Comment by maximedenes** [comment]
The plan is to have both:
- Jenkins for heavy post-integration testing, and performance regression tracking
- Travis for more lightweight pre-integration testing

**Comment by Zimmi48** [comment]
@ejgallego I think you exaggerate. I have been present to all the recent WG and have also quickly browsed the latest plannings (available there https://coq.inria.fr/cocorico/CoqDevelopment) and I have found no trace of this information. It may very well have been discussed many times but in small groups then. In any case, I don't think I am to blame for not recalling this plan.

As for Jenkins, anyone can create an account. You need to go there https://ci.inria.fr/users/new. Then you go to https://ci.inria.fr/coq/ and ask to join. Then all of us who already have an account will receive an email and anyone can allow you in. Then there is some way of adding your GitHub repo to the list of available repositories and you can test your own branch. I think @matej-kosik has some detailed explanations somewhere.

**Comment by silene** [comment]
@strub I said "more recent". Precise is 5 years old, no wonder it does not know about opam. I was thinking about xenial, for instance.

**Comment by JasonGross** [comment]
As far as I know, travis only supports trusty and precise.

**Comment by strub** [comment]
@silene, yes but precise is the last available on Travis. So this even obeys the "most recent" :)

**Comment by aspiwack** [comment]
Trusty is more recent than precise actually. That's what @silene was pointing to. Xenial is the latest LTS, but unfortunately neither Travis nor CircleCI seem to have adopted it.

That being said, if if using Xenial is clearly better, there is an easy solution: use the docker support from Travis an run a Xenial image (and do all the tests there).

**Comment by strub** [comment]
Ah yes, i should learn my alphabet. Docker is a solution, but the idea was
to have a simple way to test PR quickly.  Not to start maintaining docker
images.

Le sam. 4 févr. 2017 à 11:07, Arnaud Spiwack <notifications@github.com> a
écrit :

**Comment by aspiwack** [comment]
You should probably cache `.opam`, by the way.

Regarding docker: I think Travis will cache your docker layers, so it should be almost free to get to the point where Coq is ready to be built. Running all of this in a Docker would probably add a bit of tedium to the script, so it's all a matter of whether the benefits outweigh the cost. In this particular case, I'm not sure it does, though.

**Comment by Zimmi48** [comment]
Shouldn't this have been merged also in 8.5 and 8.6 as Matej suggested?
When going to https://travis-ci.org/coq/coq/branches we can see tests for all active branches but they are failing for 8.5 / 8.6 for lack of `.travis.yml`

**Comment by strub** [comment]
By default, travis runs on all branches, regardless of the presence of a
.travis.yml or not. Somebody should, in the settings of the Coq project on
travis, disable this behaviour --- regardless of pushing the .travis.yml in
the 8.5/8.6 branches.

2017-02-07 13:19 GMT+01:00 Théo Zimmermann <notifications@github.com>:

**Comment by matej-kosik** [comment]
@Zimmi48 I have no idea. It is like, let's create (instead of let's avoid) git-conflicts when we can.

**Comment by maximedenes** [comment]
@strub It seems the current settings do require the presence of a `.travis.yml` file, so I'm a bit confused.

@Zimmi48 As for stable branches, I wanted to experiment with trunk because we were about to merge a number of pull requests, but of course, if the experiment is successful, we will want to backport it to stable branches. Not that it does not matter so much that we are not merging forward, because most of the configuration files will have to be adjusted anyway (to point to the right upstream branch of developments to be tested).

As a final note, we did not remove any of the Jenkins infrastructure, so we can still use that to test other branches meanwhile.

**Comment by ejgallego** [comment]
@maximedenes likely these are old builds that were done before the "Require .travis.yml" setting was switched on. Travis doesn't seem to provide a way to remove builds from the web, so here you go.

**Comment by Zimmi48** [comment]
@maximedenes Is the recent frenzy of closing/reopening PRs a way to  launch Travis builds for PRs that predate the inclusion of `.travis.yml`? I initially thought it was some kind of bug you were suffering from ^^
